### PR TITLE
Adds SEO

### PR DESCRIPTION
**Pull request by Raahul-Singh**
We have a working Search Events Object!

### CODE REVIEW

**Comment by Raahul-Singh** [comment]
The codestyle test cannot be rectified, as it is only an alphabetical order shift of imports. But changing the order of imports causes other test fails.

File: pythia/seo/sunspotter.py [hunk 1]
```diff
@@ -1,11 +1,13 @@
...
+from datetime import datetime, timedelta
 from pathlib import Path
 
 import astropy.units as u
 import matplotlib.pyplot as plt
 import pandas as pd
-from astropy.coordinates import SkyCoord
+from astropy.coordinates import Longitude, SkyCoord
 from pythia.cleaning import MidnightRotation
+from pythia.seo import TableMatcher
```
Comment (dpshelio) [inline pythia/seo/sunspotter.py:10]: following the conventions of imports, imports from the same library should be in a separate block, so you've got: standard library, external dependencies, same library.

File: pythia/seo/sunspotter.py [hunk 2]
```diff
@@ -750,17 +752,22 @@ def rotate_to_midnight(self, obsdate: str, fmt='%Y-%m-%d %H:%M:%S', **kwargs):
...
         rotated = []
-        for lat in latitude:
-            rotated.append((rotator.get_longitude_at_nearest_midnight(obsdate, lat), lat))
+        try:
+            for index, lon in enumerate(longitude):
+                rotated.append((lon + Longitude(rotator.get_longitude_at_nearest_midnight(obsdate, latitude[index])),
+                                latitude[index]))
+        except TypeError:
+            rotated.append((longitude + Longitude(rotator.get_longitude_at_nearest_midnight(obsdate, latitude)),
+                                latitude))
```
Comment (dpshelio) [inline pythia/seo/sunspotter.py]: I don't understand why this `try` `except` block needs to be. Is it to be able to use one or multiple inputs? can't you use a `zip(longitude, latitude)` for the for loop instead?
```python
for lon, lat in zip(longitude, latitude):
   rotated.append((lon + Longitude(rotator.get_longitude_at_nearest_midnight(obsdate, lat)), lat))

```

File: pythia/seo/sunspotter.py [hunk 3]
```diff
@@ -858,3 +864,94 @@ def get_lat_lon_in_hgs(self, obsdate, get_nearest=True):
...
+        12964                  10843     10843             -7.105894             -2            12.301117            12
+        12965                  10844     10844             76.811298             81           -14.756412           -14
+        """
+        obsdate = self.get_nearest_observation(obsdate)
+
+        a = hek.attrs
+        client = hek.HEKClient()
+
+        tstart = datetime.strptime(obsdate, fmt)
+        tend = datetime.strptime(obsdate, fmt) + timedelta(days=days_delta)
```
Comment (dpshelio) [inline pythia/seo/sunspotter.py]:
```suggestion
        tend = tstart + timedelta(days=days_delta)
```

File: pythia/seo/sunspotter.py [hunk 4]
```diff
@@ -858,3 +864,94 @@ def get_lat_lon_in_hgs(self, obsdate, get_nearest=True):
...
+        properties = self.get_all_properties_from_obsdate(obsdate)
+
+        if isinstance(properties, pd.Series):
+            properties = pd.DataFrame(data=[properties.values],
+                                      columns=properties.index.values)
+
+        rotated = self.rotate_to_midnight(obsdate)
+
+        lon = [coord[0].value for coord in rotated]
+        lat = [coord[1].value for coord in rotated]
```
Comment (dpshelio) [inline pythia/seo/sunspotter.py]: Maybe this works...
```suggestion
        lon, lat = zip(*rotated)
```

File: pythia/seo/sunspotter.py [hunk 5]
```diff
@@ -858,3 +864,94 @@ def get_lat_lon_in_hgs(self, obsdate, get_nearest=True):
...
+
+        df_1 = pd.DataFrame(columns=['lon', 'lat'])
+        df_1['lon'] = lon
+        df_1['lat'] = lat
```
Comment (dpshelio) [inline pythia/seo/sunspotter.py]: Maybe we don't need the intermediate `lon`, `lat`
```suggestion
        df_1 = pd.DataFrame(rotated, columns=['lon', 'lat'])
```

File: pythia/seo/sunspotter.py [hunk 6]
```diff
@@ -858,3 +864,94 @@ def get_lat_lon_in_hgs(self, obsdate, get_nearest=True):
...
+
+        df_2 = data[['hgs_x', 'hgs_y']]
+
+        tablematcher = TableMatcher(match_type=match_type)
+        res = tablematcher.match(df_1, df_2, threshold=match_threshold)
```
Comment (dpshelio) [inline pythia/seo/sunspotter.py:940]: :thinking: so we are not telling here which columns to match, right? so they need to be in the same order... am I right?

Comment (dpshelio) [reply pythia/seo/sunspotter.py:940]: Oh, yeah, I got it there now :)

File: pythia/seo/tablematcher.py [hunk 7]
```diff
@@ -0,0 +1,202 @@
...
+            The row matching algorithm, by default 'cosine'
+
+        Raises
+        ------
+        SunpyUserWarning
+            If unrecognized match type is passed.
+        """
+        self.match_type = match_type
+
+        if self.match_type is not 'cosine' and self.match_type != 'euclidean':
```
Comment (dpshelio) [inline pythia/seo/tablematcher.py]: `is not` and `!=` in the same comparisson! What about:
```suggestion
        if self.match_type not in ['cosine', 'euclidean']:
```

File: pythia/seo/tablematcher.py [hunk 8]
```diff
@@ -0,0 +1,202 @@
...
+        """
+        if self.match_type == 'euclidean':
+            for index, score_value in enumerate(match_score):
+                if score_value > threshold:
+                    warnings.warn(SunpyUserWarning(f"\nMatch at Index {index} is likely to be incorrect\n"))
+
+        if self.match_type == 'cosine':
+            for index, score_value in enumerate(match_score):
+                if score_value < threshold:
+                    warnings.warn(SunpyUserWarning(f"\nMatch at Index {index} is likely to be incorrect\n"))
```
Comment (dpshelio) [inline pythia/seo/tablematcher.py]: All the above is almost equal for both. Think on a way to make it DRY-ier. A dictionary may come handy here ;)

File: pythia/seo/tablematcher.py [hunk 9]
```diff
@@ -0,0 +1,202 @@
...
+        result: `numpy.ndarray`
+            Array of size `(n,)` where n is the number of rows in df_1.
+            Contains indices of rows from df_2 that best correspond to rows from df_1.
+        """
+        df_1, df_2 = self._prepare_tables(df_1, df_2, feature_1, feature_2)
+
+        if self.match_type == 'cosine':
+            result, match_score = self.match_cosine(df_1, df_2)
+        elif self.match_type == 'euclidean':
+            result, match_score = self.match_euclidean(df_1, df_2)
```
Comment (dpshelio) [inline pythia/seo/tablematcher.py]: And a dictionary will come handy here to make this DRY-er too :wink:

File: pythia/seo/tests/test_sunspotter.py [hunk 10]
```diff
@@ -68,6 +68,21 @@ def obsdate():
     return '2000-01-01 12:47:02'
 
 
+@pytest.fixture
+def compare_df():
```
Comment (dpshelio) [inline pythia/seo/tests/test_sunspotter.py]: The name of this function is a lie :stuck_out_tongue_closed_eyes:

File: pythia/seo/tests/test_sunspotter.py [hunk 11]
```diff
@@ -275,3 +290,21 @@ def test_hpc_to_hgs_position(sunspotter, obsdate):
...
     assert isinstance(sunspotter.get_lat_lon_in_hgs(obsdate)[0], Longitude)
     assert isinstance(sunspotter.get_lat_lon_in_hgs(obsdate)[1], Latitude)
+
+
+def test_match_with_swpc_for_obsdate_default(sunspotter, compare_df):
+    obsdate = '2005-12-31 12:48:02'
+    match_result = sunspotter.match_with_swpc_for_obsdate(obsdate)
+
+    for column in compare_df.columns:
+        assert all(np.isclose(match_result[column], compare_df[column]))
```
Comment (dpshelio) [inline pythia/seo/tests/test_sunspotter.py]: I'm not too fun of having tests with asserts in a loop. If there's something wrong you will have to run it n times to get all the errors. Each time, run, find, fix, run, find, fix. Whereas running and getting all the errors at once is more handy.

File: pythia/seo/tests/test_tablematcher.py [hunk 12]
```diff
@@ -0,0 +1,117 @@
...
+                    727., 413., 312., 354., 303., 704., 606.],
+        'feat_yy': [329.343994, -644.799988, 162.688004, -656.703979, 392.832001,
+                    -319.424011, -75.391998, -668.607971, -809.471985, 577.343994,
+                    -716.223999, -242.048004, 343.231995, 599.16803, 676.544006,
+                    255.936005, 255.936005, 327.359985],
+        'feat_zz': [109.120003, -650.752014, 238.080002, -123.008003, -458.303986,
+                    -636.864014, 561.471985, -351.167999, -119.040001, 248.,
+                    345.216003, 426.559998, -196.416, -396.799988, -313.471985,
+                    -414.656006, 380.928009, 186.496002]
+	        }
```
Comment (dpshelio) [inline pythia/seo/tests/test_tablematcher.py]: Are you mixing tabs with spaces???!! :scream_cat:

File: pythia/seo/tests/test_tablematcher.py [hunk 13]
```diff
@@ -0,0 +1,117 @@
...
+
+	with pytest.warns(SunpyUserWarning):
+		matcher.match(df_1, df_2, feature_1, feature_2, threshold=0)
+
+
+def test_match_no_features_mismatch(df_1, df_2, best_match):
+    matcher = TableMatcher(match_type='euclidean')
+
+    with pytest.raises(SunpyUserWarning):
+        matcher.match(df_1, df_2, threshold=10)
```
Comment (dpshelio) [inline pythia/seo/tests/test_tablematcher.py]: Most of the test above have the same shape, can you parametrize them?

File: requirements.txt [hunk 14]
```diff
@@ -7,3 +7,4 @@ drms>=0.5.7
 zeep>=3.4.0
 matplotlib>=3.2.1
 astropy>=4.0
+scikit-learn>=0.23.1
```
Comment (dpshelio) [inline requirements.txt:10]: why do we have a `requirements.txt` file? we shouldn't need it. That should live in the setup configuration. Requirements file are a thing for websites applications :stuck_out_tongue_closed_eyes:

**Comment by dpshelio** [comment]
> The codestyle test cannot be rectified, as it is only an alphabetical order shift of imports. But changing the order of imports causes other test fails.

Why does the order of imports make things to fail?

File: pythia/seo/sunspotter.py (hunk 1 above)
Comment (Raahul-Singh) [reply pythia/seo/sunspotter.py:10]: This causes `isort` codestyle fails.

File: pythia/seo/tablematcher.py [hunk 15]
```diff
@@ -0,0 +1,201 @@
...
+            Array of size `(n,)` where n is the number of rows in df_1.
+            Contains match score for  corresponding best matches.
+        """
+        try:
+            from sklearn.metrics.pairwise import euclidean_distances
+        except ImportError:
+            raise SunpyUserWarning(
+                "Table Matcher requires Scikit Learn to be installed")
+
+        euclidean = euclidean_distances(X=df_1, Y=df_2)
```
Comment (dpshelio) [inline pythia/seo/tablematcher.py:145]: from the tests, `df_1` and `df_2` can't be compared, one has units and the other hasn't. So, somewhere we need to make sure both share the same units.

**Pull request merged**